git-worktree/src/lib.rs

```diff
@@ -1 +1,95 @@
 #![forbid(unsafe_code, rust_2018_idioms)]
+//! Git Worktree
+
+use anyhow::Result;
+use git_hash::oid;
+use git_index::{
+    entry::{Flags, Mode},
+    State,
+};
+use git_object::bstr::ByteSlice;
+use git_object::Data;
+use std::convert::TryFrom;
+use std::fs;
+use std::fs::create_dir_all;
+use std::path::Path;
+use std::time::SystemTime;
+
+#[cfg(unix)]
+use std::os::unix::fs::PermissionsExt;
+
+/// Copy index to `path`
+pub fn copy_index<P, Find>(state: &mut State, path: P, mut find: Find, opts: Options) -> Result<()>
+where
+    P: AsRef<Path>,
+    Find: for<'a> FnMut(&oid, &'a mut Vec<u8>) -> Option<Data<'a>>,
+{
+    let path = path.as_ref();
+    let paths = state
+        .entries()
+        .iter()
+        .map(|e| e.path(state).to_path().map(|p| p.to_path_buf()))
+        .collect::<Result<Vec<_>, _>>()?;
+    let mut buf = Vec::new();
+    for (i, entry) in state.entries_mut().iter_mut().enumerate() {
+        if entry.flags.contains(Flags::SKIP_WORKTREE) {
+            continue;
+        }
+        let time = SystemTime::now().duration_since(SystemTime::UNIX_EPOCH)?;
+        entry.stat.mtime.secs = u32::try_from(time.as_secs())?;
+        entry.stat.mtime.nsecs = time.subsec_nanos();
+        entry.stat.ctime.secs = u32::try_from(time.as_secs())?;
+        entry.stat.ctime.nsecs = time.subsec_nanos();
+        let dest = path.join(&paths[i]);
+        create_dir_all(dest.parent().expect("path is empty"))?;
+        match entry.mode {
+            Mode::FILE | Mode::FILE_EXECUTABLE => {
+                let obj = find(&entry.id, &mut buf).unwrap();
+                std::fs::write(&dest, obj.data)?;
+                if entry.mode == Mode::FILE_EXECUTABLE {
+                    #[cfg(unix)]
+                    fs::set_permissions(dest, fs::Permissions::from_mode(0o777)).unwrap();
+                }
+            }
+            Mode::SYMLINK => {
+                let obj = find(&entry.id, &mut buf).unwrap();
+                let linked_to = obj.data.to_path()?;
+                if opts.symlinks {
+                    #[cfg(unix)]
+                    std::os::unix::fs::symlink(linked_to, dest)?;
+                    #[cfg(windows)]
+                    if dest.exists() {
+                        if dest.is_file() {
+                            std::os::windows::fs::symlink_file(linked_to, dest)?;
+                        } else {
+                            std::os::windows::fs::symlink_dir(linked_to, dest)?;
+                        }
+                    }
+                } else {
+                    let linked_to_path = path.join(linked_to);
+                    if linked_to_path.exists() {
+                        std::fs::copy(linked_to_path, dest)?;
```

Comment:
I don't think that's how git handles symlinks, as for this would make it look like the content of the file changed. Instead, I think it will write the `linked_to` portion directly into a file - after all, that's all symlinks are on linux, which makes sure the hash matches.